### PR TITLE
Fix #90: Restrict char set of prof attributes

### LINKED ISSUE #90: Restrict character set of profiling attributes
In the not-so-distant past, we noticed that we had accumulated a lot of profiling attributes that were separated with `,` (bad) instead of `;` (good). I would suggest to check this via a r:

* GeekoDoc v1 (less strict): ` *([\-_.0-9A-Za-z]+ *; *)*([\-_.0-9A-Za-z]+) *`
* GeekoDoc v2 (more strict): `([\-0-9a-z]+;)*([\-0-9a-z]+)`

I am a bit sure if the less strict version makes enough sense -- it's likely that we could just go ahead with the stricter version and not see any fallout.

### PR DESCRIPTION
**Pull request by tomschr**
For GeekoDoc version 1 and 2

Profiling attributes (like os, arch, ...) only make sense if they are separated by semicolon. Spaces or commas are not allowed and would make the profiling algorithm to fail. Additionally, only lowercase letters, underscores, digits, and dots are allowed.

This commit introduces a regex to cover this use case.

### CODE REVIEW

File: geekodoc/rng/1_5.1/geekodoc-v1.rnc [hunk 1]
```diff
@@ -214,6 +214,16 @@ div {
   #  | SVG | PDF | SWF
   #  | linespecific
 }
+
+div {
+  # Issue #90: Restrict character set of profiling attributes
+  db.profiling.pattern =
+    xsd:string {
+      pattern = "[ ]*([\-\._0-9a-zA-Z]+[ ]*;[ ]*)*([\-\._0-9a-zA-Z]+)[ ]*"
```
Comment (sknorr) [inline geekodoc/rng/1_5.1/geekodoc-v1.rnc:222]: I looked at the code that does profiling, and afaik see (see bottom), it does _not_ apply normalize-space(). Hence, it turns out that allowing spaces anywhere, even in GD v1 is a terrible idea. Sorry, I should have checked before making that suggestion.

```suggestion
      pattern = "([\-\._0-9a-zA-Z]+;)*([\-\._0-9a-zA-Z]+)"
```

----

cf. /usr/share/xml/docbook/stylesheet/nwalsh5/current/profiling/profile-mode.xsl:235:
```xsl
<!-- Returns non-empty string if list in $b contains one ore more values from list $a -->
<xsl:template name="cross.compare">
  <xsl:param name="a"/>
  <xsl:param name="b"/>
  <xsl:param name="sep" select="$profile.separator"/>
  <xsl:variable name="head" select="substring-before(concat($a, $sep), $sep)"/>
  <xsl:variable name="tail" select="substring-after($a, $sep)"/>
<!-- <xsl:message> -->
<!-- a="<xsl:value-of select="$a"/>" -->
<!-- a="<xsl:value-of select="normalize-space($a)"/>" -->
<!-- head="<xsl:value-of select="$head"/>" -->
<!-- tail="<xsl:value-of select="$tail"/>" -->
<!-- </xsl:message> -->
  <xsl:if test="contains(concat($sep, $b, $sep), concat($sep, $head, $sep)) or normalize-space($a) = '' ">1</xsl:if>
  <xsl:if test="$tail">
    <xsl:call-template name="cross.compare">
      <xsl:with-param name="a" select="$tail"/>
      <xsl:with-param name="b" select="$b"/>
    </xsl:call-template>
  </xsl:if>
</xsl:template>
```

File: geekodoc/rng/1_5.1/geekodoc-v1.rnc [hunk 2]
```diff
@@ -1114,6 +1124,41 @@ include "docbookxi.rnc"
 
 
   #========= Changed or Redefinied Attributes
```
Comment (sknorr) [inline geekodoc/rng/1_5.1/geekodoc-v1.rnc:1126]: Typo fix while we're at it:

```suggestion
  #========= Changed or Redefined Attributes
```

**Review by sknorr — COMMENTED** [review]
Thank you!

File: geekodoc/rng/1_5.1/geekodoc-v1.rnc (hunk 1 above)
Comment (tomschr) [reply geekodoc/rng/1_5.1/geekodoc-v1.rnc:222]: Thanks for the investigation! :+1: 

I've tried your regex and I think we need to clarify one issue. Used Python to investigate but I assume, it's the same in RNG:

```python
>>> import re
>>> regex = r"([\-\._0-9a-zA-Z]+;)*([\-\._0-9a-zA-Z]+)"
# (1)
>>> re.match(regex, "")

# (2)
>>> re.match(regex, ";")

# (3)
>>> re.match(regex, "a;")
<re.Match object; span=(0, 1), match='a'>
```

Empty strings and a semicolon don't match (1 + 2). That's okay and what we want. However, a trailing semicolon matches (3). Do we really want that? Maybe the profiling stylesheets are more forgiving, I haven't checked.

If we move the semicolon a bit we could avoid that:

```python
>>> regex2 = r"([\-\._0-9a-zA-Z]+)*(;[\-\._0-9a-zA-Z]+)"

>>> re.match(regex2, "a;")
>>> re.match(regex2, "a;b")
<re.Match object; span=(0, 3), match='a;b'>
```

What do you think?

Comment (sknorr) [reply geekodoc/rng/1_5.1/geekodoc-v1.rnc:222]: It's actually not the trailing semicolon that matches -- the only thing that matches is the `a` -- i.e. there's a partial match. I am not sure exactly whether Python and RNG behavior matches here. However, you can make sure and use `^`/`$` to be sure to have a full match:

```diff
-       pattern = "[ ]*([\-\._0-9a-zA-Z]+[ ]*;[ ]*)*([\-\._0-9a-zA-Z]+)[ ]*"
+      pattern = "^([\-\._0-9a-zA-Z]+;)*([\-\._0-9a-zA-Z]+)$"
```

I think the regex syntax and behavior are imported from XML Schema though, and https://www.w3.org/TR/xmlschema11-2/#regexs explicitly notes that XML Schema behaves differently than Python here:

> Unlike some popular regular expression languages (including those defined by Perl and standard Unix utilities), the regular expression language defined here implicitly anchors all regular expressions at the head and tail

In conclusion, I think my original suggestion is probarbly fine, but you should test it:

```suggestion
      pattern = "([\-\._0-9a-zA-Z]+;)*([\-\._0-9a-zA-Z]+)"
```

**Comment by sknorr** [comment]
I've taken the liberty of merging this. Thanks for implementing it!
